### ISSUE TITLE
BannedApiAnalyzers exclude_generated_code does not suppress RS0030 for Microsoft.AspNetCore.OpenApi source-generated syntax trees

### ISSUE THREAD

**Issue by mmigala** [dotnet/roslyn#84321]
**Version Used**:

`Microsoft.CodeAnalysis.BannedApiAnalyzers`:

```xml
<PackageReference Include="Microsoft.CodeAnalysis.BannedApiAnalyzers"
                  Version="3.11.0-beta1.26226.5"
                  PrivateAssets="all" />
```

Target framework:

```xml
<TargetFramework>net10.0</TargetFramework>
```

Also using:

```xml
<PackageReference Include="Microsoft.AspNetCore.OpenApi" Version="10.0.0-preview.*" />
```

**Steps to Reproduce**:

1. Create a project targeting `net10.0` that references `Microsoft.AspNetCore.OpenApi`.

2. Add `Microsoft.CodeAnalysis.BannedApiAnalyzers` version `3.11.0-beta1.26226.5`.

3. Add a `BannedSymbols.txt` that bans APIs used by the ASP.NET Core OpenAPI source generator. In our case the generated file reports violations for APIs such as:

```text
M:System.String.Contains(System.String)
M:System.String.Replace(System.String,System.String)
```

4. Add analyzer config intended to exclude generated code:

```ini
# .globalconfig
is_global = true

dotnet_banned_api_analyzer.exclude_generated_code = true
```

5. Build the project so that `Microsoft.AspNetCore.OpenApi.SourceGenerators` emits OpenAPI generated code.

The diagnostics are reported from a generated file under `obj`, for example:

```text
obj\Debug\net10.0\Microsoft.AspNetCore.OpenApi.SourceGenerators\...\OpenApiXmlCommentSupport.generated.cs
```

We also tried explicitly marking the generated file as generated code via `.editorconfig`:

```ini
[**/obj/**/Microsoft.AspNetCore.OpenApi.SourceGenerators/**/OpenApiXmlCommentSupport.generated.cs]
generated_code = true
```

and also tried a direct severity override:

```ini
[**/obj/**/Microsoft.AspNetCore.OpenApi.SourceGenerators/**/OpenApiXmlCommentSupport.generated.cs]
dotnet_diagnostic.RS0030.severity = none
```

The diagnostic is still reported from the source-generated file.

**Diagnostic Id**:

`RS0030: Do not use banned APIs`

The diagnostic is reported against source-generated code emitted by `Microsoft.AspNetCore.OpenApi.SourceGenerators`, specifically in:

```text
OpenApiXmlCommentSupport.generated.cs
```

Example banned APIs reported from the generated file:

```text
System.String.Contains(System.String)
System.String.Replace(System.String,System.String)
```

**Expected Behavior**:

When this option is set:

```ini
is_global = true
dotnet_banned_api_analyzer.exclude_generated_code = true
```

`BannedApiAnalyzers` should not report `RS0030` diagnostics from source-generated syntax trees.

Alternatively, if a generated source file is explicitly marked as generated code using `.editorconfig`:

```ini
[**/obj/**/Microsoft.AspNetCore.OpenApi.SourceGenerators/**/OpenApiXmlCommentSupport.generated.cs]
generated_code = true
```

then `RS0030` should not be reported for that generated file.

The analyzer should continue reporting `RS0030` for handwritten source code.

**Actual Behavior**:

`RS0030` is still reported from the source-generated file:

```text
obj\Debug\net10.0\Microsoft.AspNetCore.OpenApi.SourceGenerators\...\OpenApiXmlCommentSupport.generated.cs
```

The following approaches do not suppress the diagnostic for the source-generated file:

```ini
dotnet_banned_api_analyzer.exclude_generated_code = true
```

```ini
generated_code = true
```

```ini
dotnet_diagnostic.RS0030.severity = none
```

The only reliable workarounds we found are broad or undesirable:

1. Add `<NoWarn>RS0030</NoWarn>` to the affected project, which also disables banned API enforcement for handwritten code.
2. Use a separate plugin-specific banned-symbol list that omits APIs used by the OpenAPI source generator.
3. Avoid the OpenAPI source generator.
4. Remove `BannedApiAnalyzers` from affected projects.

This makes it difficult to enforce banned APIs in handwritten code while ignoring Microsoft-generated source-generator output.

**Comment by jjonescz** [comment]
> Version="3.11.0-beta1.26226.5"

You seem to be using a very old version. Can you try https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet-tools/NuGet/Microsoft.CodeAnalysis.BannedApiAnalyzers/overview/5.10.0-1.26329.5, for example?

**Comment by mmigala** [comment]
Upgraded to 5.6.0 but still seeing the same issue

**Comment by jjonescz** [comment]
5.6.0 should have the change (https://github.com/dotnet/roslyn/pull/82713), so this looks like a real bug

**Comment by 333fred** [comment]
@mmigala I am unable to reproduce the issue on 5.6.0. Can you please provide a binlog and [complog](https://github.com/jaredpar/complog#creating-compiler-logs) of the scenario? If you cannot share it publicly, you can open a ticket on vsfeedback, upload there, and link the ticket on this issue so we can take a look.
